File: Week 6/Go Solutions/isBinaryTreeHard.go
package main

import (
  "fmt"
)

type Node struct {
  key     int
  left    int
  right   int
}

func initialize(k, l, r) *Node {
  return &Node{key: k, left: l, right: r}
}

type BSTStack struct {
  min   int
  tree  *Node
  max   int
}

func isBinary(tree []*Node) bool {
  stack := [][]BSTStack{{-999999, tree[0], 999999}}
  for len(stack) != 0 {
    min, root, max := stack[len(stack) - 1][0], stack[len(stack) - 1][1], stack[len(stack) - 1][2]
    stack = stack[0:len(stack) - 1]

    if root.key < min || root.key >= max {
      return false
    }
    if root.left != -1 {stack = append(stack, []BSTStack{min, tree[root.key], root.key})}
    if root.right != -1 {stack = append(stack, []BSTStack{root.key, tree[root.right], max})}
  }
  return true
}

func main() {
  var(
    numNodes, k, l, r    int
    nodes                []*Node
  )
  fmt.Scanln(&numNodes)

  for i := 0; i < numNodes; i++ {
    fmt.Scan(&k) ; fmt.Scan(&l) ; fmt.Scan(&r)
    nodes = append(nodes, initialize(k, l, r))
  }

  if numNodes == 0 || isBinary(nodes) {
    fmt.Println("CORRECT")
  } else {
    fmt.Println("INCORRECT")
  }
}

File: Week 6/isBinaryTreeHard.rb
# Solution passes!!!

class Node
  def initialize(key, left, right)
    @key = key
    @left = left
    @right = right
  end

  attr_reader :key, :left, :right
end

def isBST?(tree)
  stack = [[-Float::INFINITY, tree[0], Float::INFINITY]]
  while !stack.empty?
    min, root, max = stack.pop
    return false if root.key < min || root.key >= max
    stack << [min, tree[root.left], root.key] if root.left != -1
    stack << [root.key, tree[root.right], max] if root.right != -1
  end
  true
end

# Get the user's input and process it:

numNodes = gets.chomp.to_i

nodes = []
numNodes.times do |i|
  key, left, right = gets.chomp.split(/ /).map(&:to_i)
  nodes << Node.new(key, left, right)
end

puts numNodes == 0 || isBST?(nodes) ? "CORRECT" : "INCORRECT"
